src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -686,38 +691,76 @@ public async Task<VersionStamp> GetSemanticVersionAsync(CancellationToken cancel
                 analyzerConfigSet ?? _lazyAnalyzerConfigOptions);
         }
 
-        private ProjectInfo.ProjectAttributes Attributes
+        internal ProjectInfo.ProjectAttributes Attributes
             => ProjectInfo.Attributes;
 
-        private ProjectState WithAttributes(ProjectInfo.ProjectAttributes attributes)
-            => With(projectInfo: ProjectInfo.With(attributes: attributes));
+        /// <summary>
+        /// Updates <see cref="ProjectInfo"/> to a newer version of attributes.
+        /// </summary>
+        private ProjectState WithNewerAttributes(ProjectInfo.ProjectAttributes attributes)
+        {
+            // version must have already been updated:
+            Debug.Assert(attributes.Version != Attributes.Version);
+
+            return With(projectInfo: ProjectInfo.With(attributes: attributes));
+        }
+
+        public ProjectState WithAttributes(ProjectInfo.ProjectAttributes attributes, bool updateDocuments)
```

Comment:
It seems some of this code is dead here -- the only place I see this can be called is SolutionCreator.UpdateProjectInfoAsync which passes false for updateDocuments. I'm also not sure I understand why we'd expose this in the first place -- if we were to update the document checksums anyways does anything break? Because we're otherwise intentionally allowing the SolutionCreator to put ourselves in a mismatched state here and trust it'll fix it, when I'm not sure we need the trust there?

If we want to rename this to something more like WithAttributesAndDoNotUpdateDocumentsDangerous() at least that makes it clear this isn't something somebody should be calling on a regular basis. (And you can drop the argument then)

Reply:
Fixed.